File: Leetcode/Algorithm/python/Add Two Numbers.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def addTwoNumbers(self, l1, l2):
        l3 = ListNode(-1)
        ptr = l3
        
        g = 0
        while l1 or l2 or g:
            a = l1.val if l1 else 0
            b = l2.val if l2 else 0
            c = a + b + g
            
            ptr.next = ListNode(c % 10)
            g = c / 10
            ptr = ptr.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        
        return l3.next

File: Leetcode/Algorithm/kotlin/Add Two Numbers.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        var p1 = l1
        var p2 = l2
        var g = 0
        var dummyHead = ListNode(-1)
        var p = dummyHead
        
        while (p1 != null || p2 != null || g != 0) {
            var u = g + (p1?.`val` ?: 0) + (p2?.`val` ?: 0)
            p.next = ListNode(u % 10)
            g = u / 10
            p1 = p1?.next
            p2 = p2?.next
            p = p.next
        }
        return dummyHead.next
    }
}
